src/middlewared/middlewared/plugins/asigra.py

```diff
@@ -0,0 +1,358 @@
+
+            # allow local access
+            with open(asigra_pg_hba_conf, "a") as f:
+                f.write("host\tall\tall\t127.0.0.0/24\ttrust")
+
+        return True
+
+    # XXX
+    # We should probably write a postgresql middleware plugin
+    # We should also pass around a persistent DB handle
+    # XXX
+    async def asigra_database_exists(self):
+        exists = False
+        con = None
+
+        try:
+            con = pg.connect("dbname='template1' user='{}'".format(self.pg_user))
+
+        except Exception as e:
+            self.logger.error("Can't connect to template1: {}".format(e))
+            return exists
+
+        cur = con.cursor()
+        try:
+            cur.execute("SELECT COUNT(*) from pg_database where datname = 'dssystem'")
+
+        except Exception as e:
+            self.logger.error("Can't query dssystem!: {}".format(e))
+            return exists
+
+        exists = False
+        try:
+            exists = (cur.fetchone()[0] != 0)
+        except Exception:
+            exists = False
+
+        cur.close()
+        con.close()
+
+        return exists
+
+    async def asigra_database_create(self):
+        con = None
+
+        try:
+            con = pg.connect("dbname='template1' user='{}'".format(self.pg_user))
+
+        except Exception as e:
+            self.logger.error("Can't connect to template1: {}".format(e))
+            return False
+
+        con.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+        cur = con.cursor()
+
+        try:
+            self.logger.debug("there is no dssystem database found in the postgres database. Creating ...")
+            cur.execute("CREATE DATABASE dssystem")
+
+        except Exception as e:
+            self.logger.error("Can't create dssystem database: {}".format(e))
+            return False
+
+        cur.close()
+        con.close()
+
+        return True
+
+    async def asigra_database_init(self):
+        con = None
+
+        try:
+            con = pg.connect("dbname='dssystem' user='{}'".format(self.pg_user))
+
+        except Exception as e:
+            self.logger.error("Can't connect to dssystem: {}".format(e))
+            return False
+
+        con.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+        cur = con.cursor()
+
+        with open(os.path.join(self.dssystem_db_path, "postgresdssystem.sql"), "r") as f:
+            try:
+                cur.execute(f.read())
+
+            except Exception as e:
+                self.logger.error("Can't init dssystem database: {}".format(e))
+                return False
+
+        with open(os.path.join(self.dssystem_db_path, "dssystem_locale_postgres.sql"), "r") as f:
+            try:
+                cur.execute(f.read())
+
+            except Exception as e:
+                self.logger.error("Can't init dssystem database: {}".format(e))
+                return False
+
+        cur.close()
+        con.close()
+
+        return True
+
+    async def get_db_number(self):
```

Comment:
Given all the blocking operations (database) this should not be a coroutine, but a normal python method.